src/highlevel/bidib_highlevel_util.c

```diff
@@ -127,6 +127,19 @@ int bidib_start_serial(const char *device, const char *config_dir, unsigned int
 			bidib_init_threads(flush_interval);
 
 			if (!bidib_lowlevel_debug_mode) {
+				#ifdef __APPLE__
+				{
+					// On macOS, the bidib_serial_port_init() function defaults
+					// to B115200 for the baudrate, which is compatible with the
+					// BiDiB master. However, if we do not reset the BiDiB master now,
+					// then ironically the detection of the baudrate does not succeed.
+					// MSG_SYS_RESET terminates communications from a previous BiDiB session.
+					unsigned char addr_stack[] = {0x00, 0x00, 0x00, 0x00};
+					bidib_buffer_message_without_data(addr_stack, MSG_SYS_RESET, 0);
```

Comment:
Hello,
Why does the detection not succeed? 

I had a simular issue some time ago and the reason was, that the receive buffer on the mac contained some old data, that were delivered before the answer from the GBMboost. Maybe you have the same issue. Have you looked at the data that is received if you do not send the MSG_SYS_RESET first? 

MSG_SYS_RESET should be prevented at connection startup because all nodes are thrown away and the nodeTable must be recreated internally in the master.

Reply:
Hi @akuhtz, 

Thanks for your suggestion. I have tried flushing the serial port with `tcflush(...)`, and by reading and discarding all the old data, but this didn't help. 

Without the MSG_SYS_RESET hack, the connection only succeeds if the BiDiB master is powered off and on between each serial session. Otherwise, the serial session only succeeds every other attempt (the BiDiB master doesn't send MSG_SYS_MAGIC until after it has been reset).

Side point: 
The MSG_SYS_RESET hack is occurring here:
https://github.com/uniba-swt/libbidib/blob/8461c043da2ec8cae5cd910c599f53e2ea5cb52b/src/highlevel/bidib_highlevel_util.c#L138

and the node tables are created a little later:
https://github.com/uniba-swt/libbidib/blob/8461c043da2ec8cae5cd910c599f53e2ea5cb52b/src/highlevel/bidib_highlevel_util.c#L146

which calls this function:
https://github.com/uniba-swt/libbidib/blob/8461c043da2ec8cae5cd910c599f53e2ea5cb52b/src/lowlevel/bidib_lowlevel_system.c#L103

Reply:
Hello everyone,

I'm currently investigating an issue where from time to time the library crashes at startup. I'm pretty sure this is also related to old data in the buffers and I'm planning to make some changes to the startup routine such that old data will be discarded. Once I have implemented this, we can verify whether @eyip002's problems were related to that, and may remove this fix. Until then, we can use this as a workaround. However, due to exams, I can only start in April with implementing the proposed changes.

Reply:
Hello @eyip002,
I have not tested anything yet, but if I look at the code in ```bidib_detect_baudrate()``` then it looks odd to me that the first attempt is made with ```1000000``` because the master works with ```115200```. 
For me it seems that the _first attempt_ to ```bidib_communication_works()``` fails on non __APPLE__ OS (...and maybe removes old data...) and it falls back to the correct baudrate with the second try. 
On __APPLE__ it fails the first time with the correct baudrate and falls back to the false baudrate ```19200``` where the data cannot be read.

Have you tried to send 2 succeeding MSG_SYS_GET_MAGIC with a little delay (10ms) in line 80ff of ```bidib_communication_works()```?

Generally if it cannot be fixed with this change, I would leave it as it is at the moment and wait for the changes by @nicolasgross .

Reply:
Hello @akuhtz, @eyip002,
the first attempt is made with 1M baud because I wanted the library to be compatible to the BiDiB specification in general and not only to our master, and the baud rate 1M was mentioned in the [BiDiB specification](http://bidib.org/bidibus/bidib_seriell_e.html).

Reply:
Dear @akuhtz and @nicolasgross,

I just tried @akuhtz's suggestion and removed the macOS-specific starting baudrate of 115200. Now the serial connection can be established on every try without my MSG_SYS_RESET hack!

Thanks @akuhtz!

https://github.com/uniba-swt/libbidib/pull/1/files/89cec0df978196c5e0e2c6d804bb2a7ae4b06d18..5120e857a2a5fa4258930ef4655f793fa4bdc691

Reply:
I'm glad you fixed it! But didn't we implement the macOS specific starting baud rate because 1M baud was not defined on macOS? I remember that we couldn't compile libbidib on macOS with 1M baud and I'm just curious how you solved this?

Reply:
Ok I understood, you are now trying 115200 twice on macOS.

Reply:
Ah I forgot about this then:
https://github.com/uniba-swt/libbidib/blob/5120e857a2a5fa4258930ef4655f793fa4bdc691/src/transmission/bidib_transmission_serial_port.c#L99-L103

In syslog, I see this (snippet):
> libbidib started
> Serial port opened
> Trying baud rate 1000000           // On macOS, this is actually B115200
> Connection to interface could not be established
> Trying baud rate 115200             // On macOS, this is the second time trying B115200
> Send to: 0x00 0x00 0x00 0x00 seq: 0 type: MSG_SYS_DISABLE (0x04) action id: 0
> Send to: 0x00 0x00 0x00 0x00 seq: 0 type: MSG_SYS_GET_MAGIC (0x01) action id: 0
> Received from: 0x00 0x00 0x00 0x00 seq: 0 type: MSG_SYS_MAGIC (0x81) action id: 0
> Connection to interface established

So on macOS, B115200 is actually tried twice. We could think about how to properly set/report the baudrate test after your exams :-)